File: Easy/14. Longest Common Prefix/official solution (3).py
# 14. Longest Common Prefix

# Runtime: 28 ms, faster than 88.38% of Python3 online submissions for Longest Common Prefix.

# Memory Usage: 14 MB, less than 6.67% of Python3 online submissions for Longest Common Prefix.


class Solution:
    # Divide and conquer
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if len(strs) == 0:
            return ""
        else:
            return self.__longestCommonPrefix(strs, 0, len(strs) - 1)

    def __longestCommonPrefix(self, strs: List[str], left: int, right: int) -> str:
        if left >= right:
            return strs[left]
        else:
            mid = left + (right - left) // 2
            lcpLeft = self.__longestCommonPrefix(strs, left, mid)
            lcpRight = self.__longestCommonPrefix(strs, mid + 1, right)
            return self.__commonPrefix(lcpLeft, lcpRight)

    def __commonPrefix(self, left: str, right: str) -> str:
        min_len = min(len(left), len(right))
        for i in range(min_len):
            if left[i] != right[i]:
                return left[:i]
        return left[:min_len]

File: Medium/300. Longest Increasing Subsequence/official solution (3).java
// 300. Longest Increasing Subsequence

// Runtime: 11 ms, faster than 58.76% of Java online submissions for Longest Increasing Subsequence.

// Memory Usage: 37.5 MB, less than 34.00% of Java online submissions for Longest Increasing Subsequence.


class Solution {
    // Dynamic Programming
    public int lengthOfLIS(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }

        int[] count = new int[nums.length];
        count[0] = 1;

        int maxAns = 1;
        for (int i = 1; i < count.length; i++) {
            int maxVal = 0;
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    maxVal = Math.max(maxVal, count[j]);
                }
            }

            count[i] = maxVal + 1;
            maxAns = Math.max(maxAns, count[i]);
        }

        return maxAns;
    }
}